blocks/bricks/recurrent/base.py

```diff
@@ -168,28 +168,31 @@ def recurrent_apply(brick, application, application_call,
                     logger.warning("unknown input {}".format(value) +
                                    unknown_scan_input)
 
-            # Ensure that all initial states are available.
-            initial_states = brick.initial_states(batch_size, as_dict=True,
-                                                  *args, **kwargs)
-            for state_name in application.states:
-                dim = brick.get_dim(state_name)
-                if state_name in kwargs:
-                    if isinstance(kwargs[state_name], NdarrayInitialization):
-                        kwargs[state_name] = tensor.alloc(
-                            kwargs[state_name].generate(brick.rng, (1, dim)),
-                            batch_size, dim)
-                    elif isinstance(kwargs[state_name], Application):
-                        kwargs[state_name] = (
-                            kwargs[state_name](state_name, batch_size,
-                                               *args, **kwargs))
-                else:
-                    try:
-                        kwargs[state_name] = initial_states[state_name]
-                    except KeyError:
-                        raise KeyError(
-                            "no initial state for '{}' of the brick {}".format(
-                                state_name, brick.name))
-            states_given = dict_subset(kwargs, application.states)
+            if len(application.states):
+                # Ensure that all initial states are available.
+                initial_states = brick.initial_states(batch_size, as_dict=True,
```

Comment:
Again, this seems to be the only line that you want to prevent from being called when `application.states` is not given. It does not seem to be necessary to make a very big `if` clauses that damages code readability (and this is already quite complex code).